src/Layers/xrRender/ParticleEffect.cpp

```diff
@@ -453,126 +453,136 @@ __forceinline void magnitude_sse(Fvector& vec, float& res)
     _mm_store_ss((float*)&res, tv);
 }
 
-void ParticleRenderStream(FVF::LIT* pv, u32 count, PAPI::Particle * particles, CParticleEffect * pPE)
+void ParticleRenderStream(FVF::LIT* pv, u32 count, PAPI::Particle* particles, CParticleEffect* pPE)
 {
-    float sina = 0.0f, cosa = 0.0f;
-    // Xottab_DUTY: changed angle to be float instead of DWORD
-    // But it must be 0xFFFFFFFF or otherwise some particles won't play
-    float angle = 0xFFFFFFFF;
+    auto processParticle = [&](u32 i, Particle& m)
+    {
+        float sina = 0.0f, cosa = 0.0f;
+        // Xottab_DUTY: changed angle to be float instead of DWORD
+        // But it must be 0xFFFFFFFF or otherwise some particles won't play
+        float angle = 0xFFFFFFFF;
```

Comment:
I think we should get rid of this magic value and make a constant like ANGLE_NO_PARTICLES = 0xFFFFFFFF. And assigning such value to the float variable is... strange. It doesn't set all bits of the variable to 1 (as you can think when looking through this code), it assigns the value '4.29497e+09' to the variable.

Reply:
These variables was `DWORD` before, I changed the type to `float`, since they was anyway casted to floats. If we set this to `0.0f` or `1.0f` or another value, then particles stop working.
This magic value `0xFFFFFFFF` is required, if you wan't to see the particles =)
So, this value is not `ANGLE_NO_PARTICLES`, but I'd say `PARTICLES_DEFAULT_ANGLE`.

Reply:
Apples and pineapples are different fruits. Don't mix them or you'll get problems...

Reply:
Nothing wrong with this per se to force a given but pattern, but on clang and GCC you will need to disable strict aliasing or use the __may_alias__ attribute as well in the cast or you may find it optimised out.

Reply:
Nothing wrong, but the code _smells_. And necessity to disable compiler checks confirms it.

Reply:
> Nothing wrong, but the code _smells_. And necessity to disable compiler checks confirms it.
Are you shure?

Reply:
It falls foul of C++ standard rules relating to type aliasing where the standard is IMO poorly defined. Accessing the memory of one type through a pointer to another type falls into undefined behaviour, so clang and GCC will basically remove the assignment. However for floating point types this kind of type punning is actually incredibly useful and indeed the underlying maths libraries for floating point operations almost all certainly make use of it. In C doing this using a union is explicitly allowed and GCC at least allows it in C++ code as well.
Personally I would either define types using the GCC/clang attributes to allow it and use those or use a union to do the assignment as both will bring it into defined behaviour for the supported compilers. If you want to use the idiomatic but IMO silly workaround way of doing this for C++ you could do `memset(&angle, 0xFF, sizeof(angle));` to give it the correct value. Some compilers will ellide this to an assignment in the actual generated assembly making the gymnasics silly, but that is where we are.

Reply:
@OmniBlade, thank you for your detailed reply. I fully agree that accessing some data via pointer of another type is a 'gray' area and it's easy to violate rules and get UB (especially when the optimization is turned on). I'm not sure if accessing floats via pointers to dwords is explicitly allowed by the standard and won't lead to such UBs. However, even if this trick is fully legal, I'd not recommend to use it in the code if you can write the code without it. The engine is not designed for executing on 'hard-embedded' devices, so saving 4 bytes can be neglected. But it's much more important to keep the code easy to read and understand. I don't think that such trick makes our code more clear and more robust, but I'm sure that using it could lead to 'magic' bugs in the future (the other developers could miss this trick and try to use this 'invalid' value as 'valid').

Reply:
I guess if other parts of the code are actually casting back to int and comparing against this magic value, then yeah, its probably just best to refactor it out and choose some other invalid float value to assign that can be directly compared against as a float.
As far as I know, only char and unsigned char are allowed by the standard to alias other types, but only Clang and GCC to my knowledge actually optimise based on it and provide none portable attributes to get around it if you want/need to.